File: max_subarray.py
#!/usr/bin/env python
# coding=utf-8
class Solution(object):
    def max_subarray(self, nums):
        x = y = 0
        sum = n = nums[0]
        for i in range(len(nums)):
            if n > 0:
                n += nums[i]
                y = i
            else:
                n = nums[i]
                x = i
            if n > sum:
                sum = n
        return sum, x, y

nums = [-2, 3, 1, -9, 4, 8, 3]
m = Solution()
sum, x, y =  m.max_subarray(nums)
print "Max sum is {}, from index {} to {}".format(sum, x, y)

File: max_subarray.go
package main

import "fmt"

func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	n := maxSubArray(nums)
	fmt.Printf("%d\n", n)
}

func maxSubArray(nums []int) int {
	sum := nums[0]
	num := nums[0]
	for i := 1; i < len(nums); i++ {
		if num < 0 {
			num = nums[i]
		} else {
			num += nums[i]
		}
		if sum < num {
			sum = num
		}
	}
	return sum
}
